src/HonzaBotner.Services/StandUpStreakService.cs

```diff
@@ -0,0 +1,119 @@
+﻿using System;
+using System.Threading.Tasks;
+using HonzaBotner.Database;
+using HonzaBotner.Services.Contract;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+using StandUpStreak = HonzaBotner.Services.Contract.Dto.StandUpStreak;
+
+namespace HonzaBotner.Services;
+
+public class StandUpStreakService : IStandUpStreakService
+{
+    private readonly HonzaBotnerDbContext _dbContext;
+    private readonly ILogger<WarningService> _logger;
+
+    public StandUpStreakService(HonzaBotnerDbContext dbContext, ILogger<WarningService> logger)
+    {
+        _dbContext = dbContext;
+        _logger = logger;
+    }
+
+    public async Task<StandUpStreak?> GetStreak(ulong userId)
+    {
+        Database.StandUpStreak? standUpStreak = await _dbContext.StandUpStreaks
+            .FirstOrDefaultAsync(streak => streak.UserId == userId);
+
+        return standUpStreak == null ? null : GetDto(standUpStreak);
+    }
+
+    public async Task UpdateStreak(ulong userId)
+    {
+        Database.StandUpStreak? streak = await _dbContext.StandUpStreaks
+            .FirstOrDefaultAsync(streak => streak.UserId == userId);
+        ;
+
+        // Create new streak for a new user
+        if (streak is null)
+        {
+            Database.StandUpStreak newStreak = new()
+            {
+                UserId = userId,
+                Freezes = 0,
+                LastDayOfStreak = DateTime.Today.AddDays(-1),
+                Streak = 1,
+                LongestStreak = 1
+            };
+
+            await _dbContext.StandUpStreaks.AddAsync(newStreak);
+
+            try
+            {
+                await _dbContext.SaveChangesAsync();
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "Couldn't add streak {@Streak}", streak);
+            }
+
+            return;
+        }
+
+        int days = (DateTime.Today.AddDays(-2) - streak.LastDayOfStreak).Days;
+
+        if (days > streak.Freezes) //Streak broken
+        {
+            streak.Freezes = 0;
+            streak.LastDayOfStreak = DateTime.Today.AddDays(-1);
+            streak.Streak = 1;
+        }
+        else //streak restored
+        {
+            streak.Freezes -= days;
+            streak.Streak++;
+            streak.LastDayOfStreak = DateTime.Today.AddDays(-1);
+
+            if (streak.Streak > streak.LongestStreak)
+            {
+                streak.LongestStreak = streak.Streak;
+            }
+
+            if (streak.Streak % 6 == 0) // freeze acquired
```

Comment:
So freeze comes every 6 days? Why 6? 😁😅 Just asking, I'm fine with it.

Reply:
`+ move to const class member.

Reply:
@stepech Because freeze comes once a week, and maybe because I'm trying (and hopelessly fail every time) to not work on Sundays.